File: Stack/4949.py
# 균형잡힌 세상
import sys

while True:
    temp = list(sys.stdin.readline().strip())
    stack = []
    if(temp == '.'):
        break

    for word in temp:
        if word == '(':
            stack.append(word)
        elif word == '[':
            stack.append(word)
        elif word == ')':
            if not stack:
                print('no')
                continue
            elif stack[len(stack)-1] == '(':
                stack.pop()
            else:
                print('no')
                continue
        elif word == ']':
            if not stack:
                print('no')
                continue
            elif stack[len(stack)-1] == '[':
                stack.pop()
            else:
                print('no')
                continue
    if not stack:
        print("yes")

File: Javascript/Stack-Queue/4949.js
const arr = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");
const open = ["(", "["];
const closed = [")", "]"];
let stack;
const yesno = [];
arr.slice(0, arr.length-1).forEach(v => {
    let isNo = false;
    stack = [];
    for (let i=0; i<v.length; i++) {
        if (open.includes(v[i])) stack.push(v[i]);
        else if (closed.includes(v[i])) {
            if (stack.pop() !== open[closed.indexOf(v[i])]) {
                yesno.push("no");
                isNo = true;
                break;
            }
        }
    }
    if (!isNo) {
        if (stack.length === 0) yesno.push("yes");
        else yesno.push("no");
    }
});
console.log(yesno.join("\n"));